### PR TITLE
Serial Send/Receive 

### PR DESCRIPTION
**Pull request by VimFreeman**
# Assignee Section
## Description
This PR contains the library that will be used to send and receive commands. 

### Steps for Testing
Upload teensy code and send commands to it over serial.

The approval from all software team leads is necessary before merging.

# Reviewer Section

Aside from local testing and the General Integration Test it is implied that static analysis should be included in the verification process.

- [ ] Local Test Performed Successfully
- [ ] [General Integration Test](https://docs.google.com/document/d/1ug0CpA1cIzURP8DDFSvCt2CEJJSwJ6Ta6B1LG_hYk6I/edit) Performed Successfully

For Pull Requests that do not include code changes, it is not required to perform the tests above.

### CODE REVIEW

File: robot/rover/internal_comms/src/Serial.cpp [hunk 1]
```diff
@@ -16,22 +17,42 @@ namespace internal_comms
...
+        delete command;
+        command = nullptr;
     }
 
-    void sendCommand(Command* command)
-    {
-        Serial.write(command->commandID);
-        Serial.write(command->rawArgsLength);
-        Serial.write(command->rawArgs, command->rawArgsLength);
+    void sendMessage(CommandCenter* commandCenter, int enablePin) {
```
Comment (VimFreeman) [inline robot/rover/internal_comms/src/Serial.cpp:39]: enablePin is for flow control. 
For testing without flow control, ignore this parameter (or set it to =1 by default)

File: test [hunk 2]
```diff
@@ -0,0 +1 @@
+sd
```
Comment (VimFreeman) [inline test:1]: thonk

File: robot/rover/internal_comms/include/CommandCenter.h [hunk 3]
```diff
@@ -2,39 +2,79 @@
...
 
 namespace internal_comms
 {
+   typedef struct {
+       uint8_t commandID; // ID of command
+       uint8_t* rawArgs; // Byte array with the bytes of the arguments
+       uint16_t rawArgsLength; // Number of bytes in the rawArgs array
+       bool isValid; // Whether the command is valid
+   } Command;
+
```
Comment (MartensCedric) [inline robot/rover/internal_comms/include/CommandCenter.h:19]: we could have

typedef struct {
      Message* message;
      bool isValid;
} Command;

To avoid repeated code here

File: robot/rover/internal_comms/include/CommandCenter.h [hunk 4]
```diff
@@ -2,39 +2,79 @@
...
+                                        const uint8_t* rawArgs, 
+                                        const uint8_t rawArgsLength) = 0;
+
+            /**
+            * Reads serial to next command
+            * and returns a command struct
+            *
+            * @return returns a command struct
+            */
+            static Command* processCommand() ;
```
Comment (MartensCedric) [inline robot/rover/internal_comms/include/CommandCenter.h:44]: why is this static?

Comment (MartensCedric) [reply robot/rover/internal_comms/include/CommandCenter.h:44]: TBH unless there's a reasonn itt should be removed to just Command* processComand();

File: robot/rover/internal_comms/include/CommandCenter.h [hunk 5]
```diff
@@ -2,39 +2,79 @@
...
+            void queueMessage(Message& message);
+
+            /**
+             * @param cmdID command ID of message to send
+             * @param argSize number of arguments in the message
+             * @param args contains raw bytes of arguments
+             * @return returns a message struct
              */
-            Command* processCommand() const;
+            Message* createMessage(int cmdID, int argSize, byte *args);
```
Comment (MartensCedric) [inline robot/rover/internal_comms/include/CommandCenter.h:61]: it should be consistent with `executeCommand` the names of the parameters are commandID and rawArgsLength. I dont care which one you change but make sure its consistent.

File: robot/rover/internal_comms/include/CommandCenter.h [hunk 6]
```diff
@@ -2,39 +2,79 @@
...
+            etl::queue<Message, 5> messageQueue;
 
         private:
-            uint16_t readArgSize() const;
+
+            /**
+            * Reads the two bytes that make up the argument length and combines
+            * them into an uint16_t
+            */
+            static uint16_t readArgSize() ;
```
Comment (MartensCedric) [inline robot/rover/internal_comms/include/CommandCenter.h:75]: pls no static

File: robot/rover/internal_comms/src/CommandCenter.cpp [hunk 7]
```diff
@@ -1,24 +1,25 @@
...
-    Command* CommandCenter::processCommand() const
+    Command* CommandCenter::processCommand()
     {
         uint8_t commandID = waitForSerial();
         uint8_t deviceSending = waitForSerial();
         uint8_t deviceReceiving = waitForSerial();
         uint16_t argumentSize = readArgSize();
 
-        uint8_t* buffer = (uint8_t*) malloc(sizeof(uint8_t) * argumentSize);
+        auto* buffer = (uint8_t*) malloc(sizeof(uint8_t) * argumentSize);
```
Comment (MartensCedric) [inline robot/rover/internal_comms/src/CommandCenter.cpp:18]: I'd suggest you keep the explicit types and keep `auto` for foreach loops. Make it easier to know what type the variables are if they are explicitly mentioned.

File: robot/rover/internal_comms/src/CommandCenter.cpp [hunk 8]
```diff
@@ -1,24 +1,25 @@
...
         uint16_t bytesRead = (uint8_t) Serial.readBytes((char*)buffer, argumentSize);
         uint8_t stopByte = waitForSerial();
 
-        Command* cmd = (Command*) malloc(sizeof(Command));
+        auto* cmd = (Command*) malloc(sizeof(Command));
```
Comment (MartensCedric) [inline robot/rover/internal_comms/src/CommandCenter.cpp:22]: https://www.youtube.com/watch?v=MMY_SUuobww

File: robot/rover/internal_comms/src/CommandCenter.cpp [hunk 9]
```diff
@@ -37,7 +38,7 @@ namespace internal_comms
         return cmd;
     }
 
-    uint16_t CommandCenter::readArgSize() const 
+    uint16_t CommandCenter::readArgSize()
```
Comment (MartensCedric) [inline robot/rover/internal_comms/src/CommandCenter.cpp:41]: if you can put waitForSerial() const then you should put this const too

File: robot/rover/internal_comms/src/Serial.cpp [hunk 10]
```diff
@@ -16,22 +17,42 @@ namespace internal_comms
...
+        if (digitalRead(enablePin)) {
```
Comment (MartensCedric) [inline robot/rover/internal_comms/src/Serial.cpp:40]: so what happens if digitalRead(enablePin) is false(?) shouldnt it queue the message?

File: test (hunk 2 above)
Comment (MartensCedric) [inline test:1]: pls remove

**Review by MartensCedric — CHANGES_REQUESTED** [review]
looks decent so far, im just confused because you never calll queueMessage lol

File: robot/rover/internal_comms/src/CommandCenter.cpp (hunk 7 above)
Comment (VimFreeman) [reply robot/rover/internal_comms/src/CommandCenter.cpp:18]: They are explicitly mentioned in the cast

File: robot/rover/internal_comms/src/CommandCenter.cpp (hunk 8 above)
Comment (VimFreeman) [reply robot/rover/internal_comms/src/CommandCenter.cpp:22]: lol

File: robot/rover/internal_comms/include/CommandCenter.h (hunk 3 above)
Comment (VimFreeman) [reply robot/rover/internal_comms/include/CommandCenter.h:19]: would the data be accessible using cmd.*?

File: robot/basestation/static/js/elements/feed/navbar.js [hunk 11]
```diff
@@ -37,10 +37,10 @@ $(document).ready(() => {
     let url = window.location.pathname
     let toggleButtonID = 'toggle-'
 
-    if (url == '/rover') {
+    if (url == '/') {
       listener = ROVER_LISTENER_TASK
       toggleButtonID += 'rover-listener-btn'
-    } else if (url == '/' || url == '/arm') {
+    } else if (url == '/arm') {
       listener = ARM_LISTENER_TASK
```
Comment (MartensCedric) [inline robot/basestation/static/js/elements/feed/navbar.js:44]: do we need these changes lol? they seem unrelated

Comment (MartensCedric) [reply robot/basestation/static/js/elements/feed/navbar.js:44]: I guess you want to sync master with comms-master(?)

Comment (VimFreeman) [reply robot/basestation/static/js/elements/feed/navbar.js:44]: I merged master into here
I guess that might have been a mistake lel

File: robot/basestation/static/js/elements/controls/rover_controls.js [hunk 12]
```diff
@@ -75,6 +75,8 @@ $(document).ready(function () {
       }
     });
 
+    console.log('twistMsg', twistMessage.linear.x)
+    console.log('twistMsg', twistMessage.angular.z)
```
Comment (MartensCedric) [inline robot/basestation/static/js/elements/controls/rover_controls.js:79]: this shouldnt be there, even in master lol. I think its peter who pushed this directly to master lol
